integration/maptest/hammer.go

```diff
@@ -0,0 +1,678 @@
+		cfg.EmitInterval = defaultEmitSeconds * time.Second
+	}
+	return &hammerState{cfg: cfg}
+}
+
+func (s *hammerState) nextKey() string {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.keyIdx++
+	return fmt.Sprintf("key-%08d", s.keyIdx)
+}
+
+func (s *hammerState) nextValue() []byte {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.valueIdx++
+	return []byte(fmt.Sprintf("value-%09d", s.valueIdx))
+}
+
+// rev returns the revision of the specified copy of the contents, or -1 if that copy doesn't exist.
+func (s *hammerState) rev(which int) int64 {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if s.contents[which].data == nil {
+		return -1
+	}
+	return s.contents[which].rev
+}
+
+func (s *hammerState) empty(which int) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return len(s.contents[which].data) == 0
+}
+
+func (s *hammerState) label() string {
+	return strconv.FormatInt(s.cfg.MapID, 10)
+}
+
+func (s *hammerState) String() string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	details := ""
+	totalReqs := 0
+	totalInvalidReqs := 0
+	totalErrs := 0
+	for _, ep := range mapEntrypoints {
+		reqCount := int(reqs.Value(s.label(), string(ep)))
+		totalReqs += reqCount
+		if s.cfg.EPBias.Bias[ep] > 0 {
+			details += fmt.Sprintf(" %s=%d/%d", ep, int(rsps.Value(s.label(), string(ep))), reqCount)
+		}
+		totalInvalidReqs += int(invalidReqs.Value(s.label(), string(ep)))
+		totalErrs += int(errs.Value(s.label(), string(ep)))
+	}
+	return fmt.Sprintf("%d: lastSMR.rev=%s ops: total=%d invalid=%d errs=%v%s", s.cfg.MapID, smrRev(s.smr[0]), totalReqs, totalInvalidReqs, totalErrs, details)
+}
+
+func (s *hammerState) pushSMR(smr *trillian.SignedMapRoot) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Shuffle earlier SMRs along.
+	for i := smrCount - 1; i > 0; i-- {
+		s.smr[i] = s.smr[i-1]
+	}
+	s.smr[0] = smr
+}
+
+func (s *hammerState) previousSMR(which int) *trillian.SignedMapRoot {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.smr[which]
+}
+
+// pickKey randomly select a key that already exists in a particular
+// copy of the map's contents. Assumes that the selected copy is non-empty.
+func (s *hammerState) pickKey(which int) []byte {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	l := len(s.contents[which].data)
+	choice := rand.Intn(l)
+	for k := range s.contents[which].data {
+		if choice == 0 {
+			return k[:]
+		}
+		choice--
+	}
+	panic("internal error: inconsistent count of map contents")
+}
+
+// pickCopy returns an index for a non-empty copy of the map's
+// contents.  If there are no non-empty copies, it returns -1, false.
+func (s *hammerState) pickCopy() (int, bool) {
+	// Count the number of filled copies.
+	i := 0
+	for ; i < copyCount; i++ {
+		if len(s.contents[i].data) == 0 {
+			break
+		}
+	}
+	if i == 0 {
+		// No copied contents yet
+		return -1, false
+	}
+	return rand.Intn(i), true
+}
+
+func (s *hammerState) updateContents(rev int64, leaves []*trillian.MapLeaf) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Shuffle earlier contents along.
```

Comment:
I wonder would using `container/list` or perhaps `container/ring` help?

Reply:
Not obviously.  `container/list` has no random access, which we use a bunch, and `container/ring` doesn't have the concept of a head, which we need for the 'most-recent' entries.